### PR TITLE
[MIG][REF] web_advanced_search: Rename, refactor, migrate

### PR DESCRIPTION
**Pull request by yajo**
* Includes https://github.com/OCA/web/pull/984
* Git history cleanup
* Complete migration to v11
* Refactor to use the new v11 decoupled widgets system
* Advanced search is now a high-level feature from the filters menu; it simplifies code a lot, and the UX is even better
* Split README system
* Add fun to ROADMAP
* Addon is renamed to web_advanced_search, since it enhaces the searching experience for all kind of fields now

@Tecnativa

### CODE REVIEW

File: web_advanced_search/__manifest__.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
+# -*- coding: utf-8 -*-
+# Copyright 2015 Therp BV <http://therp.nl>
+# Copyright 2017 Tecnativa - Vicent Cubells
```
Comment (pedrobaeza) [inline web_advanced_search/__manifest__.py]: Add your copyright here

File: web_advanced_search/static/src/js/web_advanced_search_x2x.js [hunk 2]
```diff
@@ -0,0 +1,283 @@
+/* Copyright 2015 Therp BV <http://therp.nl>
```
Comment (pedrobaeza) [inline web_advanced_search/static/src/js/web_advanced_search_x2x.js]: Rename file according module renaming

File: web_advanced_search/__init__.py [hunk 3]
```diff
@@ -0,0 +1,3 @@
+# -*- coding: utf-8 -*-
```
Comment (pedrobaeza) [inline web_advanced_search/__init__.py]: Remove coding

File: web_advanced_search/__manifest__.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
+# -*- coding: utf-8 -*-
```
Comment (pedrobaeza) [inline web_advanced_search/__manifest__.py]: Remove coding

**Review by pedrobaeza — DISMISSED** [review]
Functionally, the only problem I find is with the resulting filter label, putting the technical domain. Isn't any way to put a text description of the filter like in previous versions happen?

File: web_advanced_search/README.rst [hunk 5]
```diff
@@ -0,0 +1,68 @@
+.. image:: https://img.shields.io/badge/license-AGPL--3-blue.svg
+   :target: http://www.gnu.org/licenses/agpl-3.0-standalone.html
+   :alt: License: AGPL-3
+
+=========================================
+Search for x2x records in advanced search
+=========================================
+
+Standard behavior in advanced search for one2many, many2many and many2one fields is to do a `name_search`. This often is not satisfactionary as you might want to search for other properties. There might also be cases where you don't exactly know what you're searching for, then a list of possible options is necessary too. This module enables you to have a full search view to select the record in question, and either select specific records or select them using a search query of its own.
```
Comment (elicoidal) [inline web_advanced_search/README.rst]: satisfactory

File: web_advanced_search/README.rst [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+Usage
+=====
+
+To use this module, you need to:
+
+* open the advanced search options in a search view
```
Comment (elicoidal) [inline web_advanced_search/README.rst]: should not this a be enumerated list?

File: web_advanced_search/README.rst [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+* select a one2many, many2many or many2one field
+* select operator `is equal to` or `is not equal to`
+* the textfield changes to a many2one selection field where you can search for the record in question
+* click *Apply*
+
+To search for properties of linked records (ie invoices for customers with a credit limit higher than X):
+
+* open the advanced search options in a search view
```
Comment (elicoidal) [inline web_advanced_search/README.rst]: and here?

File: web_advanced_search/README.rst [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+Note that you can stack searching for properties: Simply add another advanced search in the selection search window. You can do this indefinetely, so it is possible to search for moves belonging to a journal which has a user who is member of a certain group etc.
+
+.. image:: https://odoo-community.org/website/image/ir.attachment/5784_f2813bd/datas
+   :alt: Try me on Runbot
+   :target: https://runbot.odoo-community.org/runbot/162/10.0
+
+Known issues / Roadmap
+======================
+
+* When you use *is in selection* search system and choose a domain, it gets
```
Comment (elicoidal) [inline web_advanced_search/README.rst]: A list of one item is not a list :wink:

File: web_advanced_search/__manifest__.py [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+    "depends": [
+        'web',
+    ],
+    "data": [
+        'views/templates.xml',
+    ],
+    "qweb": [
+        'static/src/xml/web_advanced_search.xml',
+    ],
+    "auto_install": False,
```
Comment (elicoidal) [inline web_advanced_search/__manifest__.py]: no need for the keys with the default values no?

File: web_advanced_search/static/src/xml/web_advanced_search.xml [hunk 10]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!-- Copyright 2017 Jairo Llopis <jairo.llopis@tecnativa.com>
```
Comment (elicoidal) [inline web_advanced_search/static/src/xml/web_advanced_search.xml]: update the date

**Review by elicoidal — APPROVED** [review]
some details.
Simple tests OK

File: web_advanced_search/README.rst (hunk 5 above)
Comment (yajo) [reply web_advanced_search/README.rst]: Please review the split README (in the `readme/` folder).

This file is untouched, waiting for it to be overwritten by the oca bot that writes README files when it founds split files.

File: web_advanced_search/README.rst (hunk 6 above)
Comment (yajo) [reply web_advanced_search/README.rst]: same here

File: web_advanced_search/README.rst (hunk 7 above)
Comment (yajo) [reply web_advanced_search/README.rst]: same here.

File: web_advanced_search/README.rst (hunk 8 above)
Comment (yajo) [reply web_advanced_search/README.rst]: same here.

**Review by yajo — COMMENTED** [review]
> Functionally, the only problem I find is with the resulting filter label, putting the technical domain. Isn't any way to put a text description of the filter like in previous versions happen?

Yes, it's in the known issues. We'd need a tool to translate a domain to a human-readable format, which should possibly be upstream, either in odoo itself, or in the module that extends the domain widget, and then used downstream in this addon.

It doesn't seem as a blocker anyway, isn't it? :innocent:

File: web_advanced_search/__manifest__.py [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+    ],
+    "data": [
+        'views/templates.xml',
+    ],
+    "qweb": [
+        'static/src/xml/web_advanced_search.xml',
+    ],
+    'installable': True,
+    "application": False,
+}
```
Comment (MeetKD) [inline web_advanced_search/__manifest__.py]: website key is missing.

**Review by MeetKD — CHANGES_REQUESTED** [review]

File: web_advanced_search/static/src/js/web_advanced_search.js [hunk 12]
```diff
@@ -0,0 +1,283 @@
...
+     * has no data on it by definition.
+     */
+    var Relational = Char.extend(FieldManagerMixin, {
+        tagName: "div",
+        className: "x2x_container",
+        attributes: {},
+
+        init: function () {
+            this._super.apply(this, arguments);
+            // To make widgets work, we need to a model and an empty record
```
Comment (simahawk) [inline web_advanced_search/static/src/js/web_advanced_search.js]: *to add

File: web_advanced_search/static/src/js/web_advanced_search.js [hunk 13]
```diff
@@ -0,0 +1,283 @@
...
+            );
+            this.propositions = [proposition];
+            this.commit_search();
+        },
+    });
+
+    /**
+     * A search field for relational fields.
+     *
+     * It implements and extends the `FieldManagerMixing`, and acts as if it
```
Comment (simahawk) [inline web_advanced_search/static/src/js/web_advanced_search.js]: *Mixin

File: web_advanced_search/static/src/js/web_advanced_search.js [hunk 14]
```diff
@@ -0,0 +1,283 @@
...
+                        case "=":
+                            return -2;
+                        case "!=":
+                            return -1;
+                        default:
+                            return 0;
+                    }
+                });
+            // Create dummy record with only the field the user is searching
+            var params = {
```
Comment (simahawk) [inline web_advanced_search/static/src/js/web_advanced_search.js]: I'd move this to a separated function

**Review by simahawk — APPROVED** [review]
code & functional review: LG.
Small things I'd change but not mandatory.

Tested on runbot: the "search more" is a bit glitchy as it closes the filter dropdown but it's something we can live with. Worth to drop a line in roadmap?

**Comment by yajo** [comment]
WIP until I include fixes for @simahawk's suggestions and I backport human-readable search string generator from https://github.com/odoo/odoo/pull/25922.

> Tested on runbot: the "search more" is a bit glitchy as it closes the filter dropdown but it's something we can live with. Worth to drop a line in roadmap?

That's something that I actually like, because when hitting "apply", you don't have to apply it again. What's the problem on that? :thinking:

File: web_advanced_search/static/src/js/web_advanced_search.js (hunk 14 above)
Comment (yajo) [reply web_advanced_search/static/src/js/web_advanced_search.js]: I have left this as it was. Thinking about it, it's only done once at init time, and calling it again could produce some problems, so it seems legit to not do a method about it IMHO.

I fixed your other comments though.

**Comment by yajo** [comment]
Nice, we have now human-readable leaf strings for advanced searches.

I was expecting https://github.com/odoo/odoo/pull/25922 to get merged faster and me being able to backport it to OCB, but whatever. It's here in the addon now.

If other addons require this system, it could be split in a separate addon, but I think this is a good approach to keep it simple for now, when that's not happening.

This is how it looks now: 
captura de pantalla de 2018-08-13 13-10-31

**Comment by pedrobaeza** [comment]
@Yajo excellent addition!

One nitpicking about the human readable format: please use `→` character instead of `>` and remove spaces for better visualization and for differentiating from the operator "greater than". Are `and` and `or` words translatable?

@simahawk can you update your review?

**Comment by yajo** [comment]
> Are and and or words translatable?

Yes.

**Review by pedrobaeza — APPROVED** [review]

**Comment by pedrobaeza** [comment]
Merged in e764017

**Comment by JBF91** [comment]
Why **__manifest__.py** became **__openerp__.py** after merging?

**Comment by yajo** [comment]
:confused: weird...

**Comment by pedrobaeza** [comment]
Mea culpa, I screwed up on the squashing. Fixing directly.

**Pull request closed without merging**